statsmodels/discrete/tests/test_truncated_model.py

```diff
@@ -0,0 +1,388 @@
+
+
+class TestZeroTruncatedPoissonModel(CheckResults):
+    @classmethod
+    def setup_class(cls):
+        data = datasets.randhie.load()
+        exog = add_constant(np.asarray(data.exog)[:, :4], prepend=False)
+        mod = TruncatedPoisson(data.endog, exog, truncation=0)
+        cls.res1 = mod.fit(maxiter=500)
+        res2 = RandHIE()
+        res2.zero_truncated_poisson()
+        cls.res2 = res2
+
+
+class TestZeroTruncatedNBPModel(CheckResults):
+    @classmethod
+    def setup_class(cls):
+        data = datasets.randhie.load()
+        exog = add_constant(np.asarray(data.exog)[:, :3], prepend=False)
+        mod = TruncatedNegativeBinomialP(data.endog, exog, truncation=0)
+        cls.res1 = mod.fit(maxiter=500)
+        res2 = RandHIE()
+        res2.zero_truncted_nbp()
+        cls.res2 = res2
+
+    def test_conf_int(self):
+        pass
+
+
+class TestTruncatedPoisson_predict(object):
+    @classmethod
+    def setup_class(cls):
+        cls.expected_params = [1, 0.5]
+        np.random.seed(123)
+        nobs = 200
+        exog = np.ones((nobs, 2))
+        exog[:nobs//2, 1] = 2
+        mu_true = exog.dot(cls.expected_params)
+        cls.endog = truncatedpoisson.rvs(mu_true, 0, size=mu_true.shape)
+        model = TruncatedPoisson(cls.endog, exog, truncation=0)
+        cls.res = model.fit(method='bfgs', maxiter=5000)
+
+    def test_mean(self):
+        assert_allclose(self.res.predict().mean(), self.endog.mean(),
+                        atol=2e-1, rtol=2e-1)
+
+    def test_var(self):
+        v = self.res.predict(which="var").mean()
+        assert_allclose(v, self.endog.var(), atol=2e-1, rtol=2e-1)
+        return
+        assert_allclose((self.res.predict().mean() *
+                        self.res._dispersion_factor.mean()),
+                        self.endog.var(), atol=5e-2, rtol=5e-2)
+
+    def test_predict_prob(self):
+        res = self.res
+
+        pr = res.predict(which='prob')
+        pr2 = truncatedpoisson.pmf(
+            np.arange(8), res.predict(which="mean-main")[:, None], 0)
+        assert_allclose(pr, pr2, rtol=1e-10, atol=1e-10)
+
+
+class TestTruncatedNBP_predict(object):
+    @classmethod
+    def setup_class(cls):
+        cls.expected_params = [1, 0.5, 0.5]
+        np.random.seed(1234)
+        nobs = 200
+        exog = np.ones((nobs, 2))
+        exog[:nobs//2, 1] = 2
+        mu_true = np.exp(exog.dot(cls.expected_params[:-1]))
+        cls.endog = truncatednegbin.rvs(
+            mu_true, cls.expected_params[-1], 2, 0, size=mu_true.shape)
+        model = TruncatedNegativeBinomialP(cls.endog, exog, truncation=0, p=2)
+        cls.res = model.fit(method='nm', maxiter=5000, maxfun=5000)
+
+    def test_mean(self):
+        assert_allclose(self.res.predict().mean(), self.endog.mean(),
+                        atol=2e-1, rtol=2e-1)
+
+    def test_var(self):
+        v = self.res.predict(which="var").mean()
+        assert_allclose(v, self.endog.var(), atol=1e-1, rtol=1e-2)
```

Comment:
using `endog.var()` is not a good unit test
this assumes mean in the var computation is constant
I used var of resid (resid_reponse) in my checks.

a better test would be that pearson_chi2 is close to 1 in a correctly specified model

same for the Poisson predict test above